src/man/man1/iclass.1
.TH iclass 1 
.SH NAME
iclass \- class browser
.SH SYNOPSIS
.B iclass
.RB [ -r ]
.RB [ -v ]
.RB [ -c ]
.I path
.RI [ path... ]
.SH DESCRIPTION
.I Iclass
is a simple class browser that lets you navigate through C++ class
hierarchies and view the definition of a particular class.  You can
select classes from three browsers: the first shows a list of all
known classes; the second shows the parents of the selected class, if
any; and the third shows the children of the selected class, if any.
You can choose a name from any of these lists and tell \fIiclass\fP to
show the definition of that class, or you can type the name of a class
explicitly in the StringEditor labeled ``selected class.''  The view
of the class definition shows the source code from the header file
that defines the selected class.
.SH ARGUMENTS
\fIIclass\fP interprets arguments that specify the header files to
search for class definitions.
Each non-option argument on the command line should specify a pathname
to a header file or a directory containing header files.
If the pathname is a filename ending in the suffix ``.h'',
then \fIiclass\fP will scan that header file.
If the pathname is a directory,
then it will scan every header file in that directory.
\fIIclass\fP will ignore filenames that do not end in the suffix ``.h'',
unless the ``\-c'' option is given.
In this case \fIiclass\fP will search both ``.h'' and ``.C'' files for
class definitions.
.PP
You may use any of the standard options accepted by InterViews
programs in addition to the following options:
.TP
.B \-r
Recursively scan subdirectories for additional header files when
scanning a directory.
.TP
.B \-v
Verbosely print the name of every header file scanned.
.TP
.B \-c
Allow \fIiclass\fP to search ``.C'' files as well as ``.h'' for class
definitions.
This option should be used carefully.
Because of the way that iclass searches for definitions, it is possible
for it to get confused by code in a ``.C'' file and think it is a definition.
Nonetheless, this is a useful option if there are C++ input files that
contain definitions.
.SH COMMANDS
\fIIclass\fP provides pull-down menus for issuing commands.  The
``File'' menu includes the following commands:
.TP 4
.B "Scan Files/Directories..."
Prompt for additional pathnames to scan.
.TP
.B "Clear Classes"
Purge iclass of all classes it has scanned.
.TP
.B Quit
Close the window and exit.
.PP
Commands in the ``Search'' menu include:
.TP
.B "Forward Search..."
Search forward in the class definition for a match with a given
regular expression, wrapping from the end to the beginning of the
buffer if necessary.
.TP
.B "Backward Search..."
Search backward for a match with a given regexp.
.TP
.B "Go to Line..."
Move the cursor to the beginning of the given line, scrolling the
display if necessary.
.SH MOUSE BINDINGS
The left mouse button selects text.  Left-clicking inside a browser,
StringEditor, or class definition also focuses keyboard input to it,
which lets it interpret subsequent keyboard commands as scrolling or
selection operations.  Double left-clicking inside a browser is
equivalent to selecting a class and then pressing the ``Show'' button.
.PP
The middle and right mouse buttons scroll text in different ways.
Middle-clicking lets you ``grab-scroll'' the text.  During
grab-scrolling, the text is scrolled to follow the mouse position,
making it appear as though you are dragging the text itself.
Right-clicking engages ``rate-scrolling,'' a joy-stick-like scrolling
interface in which the scrolling rate increases as you drag the mouse
away from the initial click point.  For example, dragging the mouse
downwards after the initial click scrolls a browser downwards at an
increasing rate; dragging upwards thereafter reduces the rate until
scrolling stops entirely at the initial click point.  Dragging up
beyond this point makes the browser scroll in the reverse direction.
.SH KEYBOARD BINDINGS
For convenience, all menu commands have keyboard equivalents:
.TP
.B "(^V)"
Scan a file or directory.
.TP
.B "(^Q)"
Quit.
.TP
.B "(^S)"
Search forward.
.TP
.B "(^R)"
Search backward.
.TP
.B "(^N)"
Go to a line.
.PP
\fIIclass\fP also provides keyboard commands for changing focus:
.TP
.B "ESC or TAB"
Set focus to the StringEditor showing the selected class.
.TP
.B "CR"
Show the selected class (equivalent to pressing the Show button) and
set focus to the class definition.
.SH SEE ALSO
Regexp(3I), StringBrowser(3I), StringEditor(3I), TextEditor(3I)


CHANGES
May 12th 2012  ivtools-1.2.10

Many practical and small extensions to comterp, the scripting language embedded in
comdraw, flipbook, graphdraw, and drawserv.

Dropped complete support for gcc 3.0 and earlier. Things still might work but you'd be better off
upgradeing to gcc 4.0 or staying with ivtools-1.2.9.

Feb. 23rd, 2010  ivtools-1.2.9

- Debian related changes suggested by Agustin Martin.

- usec timeout on comdraw pause command -- allows for writing loops
  that can be exited.

- option for an end-of-line token introduced to LexScan.

- port to 64 bit Fedora

- fix problem with negative numbers in attribute editor

- make File/Save and File/Revert always work regardless of modified flag

Nov. 21st, 2009  ivtools-1.2.8

- significant work on keeping up with gcc and libstdc++.  A wrapper
  for stdio.h was inadvertently getting pulled in by /usr/include
  files, causing a lot of trouble.  Now builds with gcc-4.4 on Ubuntu
  9.10.

August 25th, 2009  ivtools-1.2.7

- series of memory leak checking commands: alistleak, attrvleak, commandleak, compleak, graphlicleak, mlineleak, viewleak.  Each returns the current number of allocated objects of that type (AttributeValueList, AttributeValue, Command, OverlayComp, Graphic, MultiLineObj, and OverlayView).

- added reference counting to OverlayView's (ComponentView's) referred from AttributeValue's (#define RESOURCE_COMPVIEW)

- fixed attrname func.

- fixed stddev func.

- added BooleanType to NumFunc promotion method.

- fixed graphdraw copy/paste of graph fragments

- new SocketObj for use comterp

- fix behavior of symbol manipulating funcs, by using internal bquote.

- port to Ubuntu Heron

- applied (most of) patch to build on Fedora 10 contributed by John Heidemann

- balance of changes to support ipl-1.0.4

July 9th, 2008  ivtools-1.2.6

- Debian contributed patch for iostreams backward compatibility

- bug fix for dispatcher.c contributed by Damon Permezel

- add mute command to comterp

- balance of changes to support ipl-1.0.2

March 4th, 2008  ivtools-1.2.5

- overall evolution to support use by ipl-1.0.1 (see
  http://sf.net/projects/ipl for further details)

July 20th, 2006  ivtools-1.2.4

- changes for compiling with gcc-4.0

- changes to socket-driven (ACE) use of comterp to support multi-line commands

- changes to allow re-use of scanner/parser with multiple operator tables

October 7th, 2005  ivtools-1.2.3

- add support for an external libtiff with a --with-tiff (and
-with-tiff-lib) configure arguments.

- changes necessary to compile with gcc-4.0

- changes necessary to compile with gcc-4.0 on top of ACE-5.4.7

- changes necessary to compile on suse-9.3 linux (amd64 platform).
Contributed by Paul Vixie.

- fixed bug in "addtool" command/mechanism in graphdraw and drawserv

- added support for color expressions to graphdraw (:colexpr), to
allow attribute-sensitive computation of a node's color.

January 25th, 2005  ivtools-1.2.2

If you submitted a patch in the past half year to ivtools that you
don't see reflected here, please resubmit.  I lost that e-mail.

- fix for problem writing to /tmp

- compilation fix on NetBSD for comterp.c

- compilation patch submitted by Michal Palczew.

July 13th, 2004  ivtools-1.2.1

Two crashes discovered and fixed by Michal Palczewski (mpalczew at
users.sourceforge.net).  One was a crash on exit after a delete with
idraw/drawtool, the other was a crash upon delete with graphdraw.
They turned out to be long-standing bugs in the order of object
deletion, problems that hadn't appeared with other versions of gcc.
RedHat must have adopted a more aggressive recycling of freed space in
a recent release.  Thanks Michal.


June 3rd, 2004  ivtools-1.2

ivtools-1.2 sets a new direction for the package, toward peer-to-peer
drawing chat.  ivtools' drawserv, the editor that incorporates every ivtool
mechanism to date (multi-frame, graph-editing, integrated scripting
language) has been extended to support a unique kind of
small-granularity shared whiteboard.  

This work was based on the pre-existing networked scripting capability
of ivtools' comdraw.  Then the pasting of any graphic was intercepted,
and the graphic serialized into a command that could be communicated
to other drawing editors to keep them in sync.

Then a connection management system was established, that supported
two-way peer-to-peer links between drawing editors.  In addition the
return path on each socket was used for detecting timeout, which
causes the connection to be dropped.

Finally, a distributed mutually-exclusive graphic selection mechanism
was created, where every drawing editor is aware of who currently has
a graphic selected or who had it selected last if it is not currently
selected.  When one drawing editor tries to select a graphic, it first
requests that permission from the last selector (after determining it
is unselected).

Now you can link to other drawing editors, draw graphics and see them
appear remotely (although only in black and white), and exercise the
distributed selection mechanism.  Four pieces remain to be developed:
1) support for full-up graphic state transmittal (it is sent, but not
interpreted on the remote end);, 2) support for transmission of the
complete set of drawing editor commands; 3) extending the distributed
mutual-exclusion mechanism to select pseudo-graphics to acquire global
permission for connection, pasting, moving front-to-back, and
grouping; and 4) a two-tier mechanism for invoking select(), that will
allow the bypass of selection requests while engaged in lengthy direct
manipulations.

If all this interests you, and you'd like to help, build this version
of ivtools with ACE support, subscribe to the ivtools-dev mailing, and
help me do some testing, programming, thinking, whatever.  Or visit
http://www.ivtools.org to make a donation to support this work.  Thanks.

Scott Johnston




src/iclass/iclass.c
/*
 * Copyright (c) 1989 Stanford University
 *
 * Permission to use, copy, modify, distribute, and sell this software and its
 * documentation for any purpose is hereby granted without fee, provided
 * that the above copyright notice appear in all copies and that both that
 * copyright notice and this permission notice appear in supporting
 * documentation, and that the name of Stanford not be used in advertising or
 * publicity pertaining to distribution of the software without specific,
 * written prior permission.  Stanford makes no representations about
 * the suitability of this software for any purpose.  It is provided "as is"
 * without express or implied warranty.
 *
 * STANFORD DISCLAIMS ALL WARRANTIES WITH REGARD TO THIS SOFTWARE,
 * INCLUDING ALL IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS.
 * IN NO EVENT SHALL STANFORD BE LIABLE FOR ANY SPECIAL, INDIRECT OR
 * CONSEQUENTIAL DAMAGES OR ANY DAMAGES WHATSOEVER RESULTING FROM LOSS OF USE,
 * DATA OR PROFITS, WHETHER IN AN ACTION OF CONTRACT, NEGLIGENCE OR
 * OTHER TORTIOUS ACTION, ARISING OUT OF OR IN CONNECTION
 * WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
 */

/*
 * IClass implementation.
 */

#include "classbuffer.h"
#include "classeditor.h"
#include "classinfo.h"
#include "dialogs.h"
#include "globals.h"
#include "iclass.h"

#include <InterViews/adjuster.h>
#include <InterViews/border.h>
#include <InterViews/box.h>
#include <InterViews/button.h>
#include <InterViews/compeditor.h>
#include <InterViews/filebrowser.h>
#include <InterViews/font.h>
#include <InterViews/frame.h>
#include <InterViews/glue.h>
#include <InterViews/menu.h>
#include <InterViews/message.h>
#include <InterViews/regexp.h>
#include <InterViews/painter.h>
#include <InterViews/scrollbar.h>
#include <InterViews/sensor.h>
#include <InterViews/strbrowser.h>
#include <InterViews/streditor.h>
#include <InterViews/textbuffer.h>
#include <InterViews/world.h>
#include <OS/math.h>
#include <OS/types.h>

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/stat.h>

/*****************************************************************************/

static const float fspace = .375;                // space in cm

static const char* NONE = "<none>";

static const int BUFFERSIZE = 100;
static const int FILENAMESIZE = 100;
static const int MINTEXTSIZE = 10000;

#define BSRCH_LBL "^R"
#define BSRCH_CODE '\022'

#define FSRCH_LBL "^S"
#define FSRCH_CODE '\023'

#define GOTO_LBL "^N"
#define GOTO_CODE '\016'

#define QUIT_LBL "^Q"
#define QUIT_CODE '\021'

#define SCAN_LBL "^V"
#define SCAN_CODE '\026'

#define CLEAR_LBL "^C"
#define CLEAR_CODE '\003'

/*****************************************************************************/

class Command : public MenuItem {
public:
    Command(const char* lbl, const char* klbl, char kcode, IClass*);

    virtual void Do();
private:
    IClass* _iclass;
    char _kcode;
};

Command::Command (
    const char* lbl, const char* klbl, char kcode, IClass* s
) : MenuItem((Interactor*) nil) {
    Insert(
       new HBox(
            new Message(lbl, Left, 2, hfil),
            new Message("   ", Center, 2, hfil),
            new Message(klbl, Right, 2, hfil)
        )
    );
    _iclass = s;
    _kcode = kcode;
}

void Command::Do () { _iclass->Command(_kcode); }

struct MenuInfo {
    const char* _lbl;
    const char* _klbl;
    char _kcode;
};

/*****************************************************************************/

IClass::IClass (ClassBuffer* cb) {
    SetClassName("IClass");
    Init(cb);
    Visit("");
    _editor->Edit(_text);
}

IClass::~IClass () {
    delete _cbuf;
    delete _buf;
    delete _text;
    delete _lastFile;
    _state->Detach(this);
    Unref(_state);

    if (_scanner != nil) delete _scanner;
    if (_fwdSearch != nil) delete _fwdSearch;
    if (_bwdSearch != nil) delete _bwdSearch;
    if (_goto != nil) delete _goto;
    if (_completions != nil) delete _completions;
}

void IClass::Init (ClassBuffer* cb) {
    _cbuf = cb;
    _state = new ButtonState;
    _state->Attach(this);
    _curClass = new CompletionEditor(_state, "");
    _focus = _curClass;
    _classes = new StringBrowser(_state, 5, 24);
    _parents = new StringBrowser(_state, 2, 24);
    _children = new StringBrowser(_state, 5, 24);
    _fileIndic = new MarginFrame(new Message("testing"));
    _editor = new ClassEditor(_state, 24, 80, 8, Reversed);
    _editor->SetScrollAlignment(TopLeft);

    _scanner = new FileDialog("Scan through files/directories:", ".");
    _fwdSearch = nil;
    _bwdSearch = nil;
    _goto = nil;

    _buf = nil;
    _text = nil;
    _lastFile = nil;
    _completions = nil;

    input = new Sensor;
    input->Catch(KeyEvent);

    Insert(Interior());
    UpdateClassesBrowser();
}

void IClass::Visit (const char* filename) {
    if (_lastFile == nil || strcmp(_lastFile, filename) != 0) {
        class Message* msg = new Message(_scanner->FullPath(filename));
        _fileIndic->Insert(msg);
        _fileIndic->Change(msg);

        delete _lastFile;
        _lastFile = strnew(filename);

        delete _buf;
        delete _text;

        FILE* f = nil;

        if (strlen(filename) != 0) {
	    /* cast to work around bug in prototype on some systems */
            f = fopen((char*)filename, "r");
        }

        if (f != nil) {
            struct stat filestats;
            stat(filename, &filestats);
            _bufsize = max(Math::round(filestats.st_size * 1.2), MINTEXTSIZE);
            _buf = new char[_bufsize];
            char* b = _buf;
            int remaining = _bufsize;
            while (remaining > 1 && fgets(b, remaining, f) != nil) {
                int l = strlen(b);
                remaining -= l;
                b += l;
            }
            fclose(f);
            _text = new TextBuffer(_buf, b-_buf, _bufsize);

        } else {
            _bufsize = MINTEXTSIZE;
            _buf = new char[_bufsize];
            _text = new TextBuffer(_buf, 0, _bufsize);
        }
    }
}

void IClass::Run () {
    Event e;
    int value;

    do {
        Read(e);
	if (e.target == nil) {
	    /* probably quit from window manager */
	    break;
	}
        Handle(e);

        _state->GetValue(value);
    } while (value != QUIT_CODE);
}

void IClass::Handle (Event& e) {
    if (e.eventType == KeyEvent && e.len > 0) {
        char c = e.keystring[0];

        if (!Command(c)) {
            _focus->Handle(e);
        }

    } else if (e.target != this && e.eventType != KeyEvent) {
        boolean focusable =
            e.target == _classes || e.target == _parents ||
            e.target == _children || e.target == _editor;

        if (focusable && e.button == LEFTMOUSE) {
            UnselectCurClass();
            UnselectBrowsers();
            _focus = e.target;
        }
        e.target->Handle(e);
    }
}

boolean IClass::Command (char c) {
    boolean executed = true;

    switch (c) {
        case BSRCH_CODE:         BackwardSearchCmd(); break;
        case FSRCH_CODE:         ForwardSearchCmd(); break;
        case GOTO_CODE:          GotoCmd(); break;
        case QUIT_CODE:          QuitCmd(); break;
        case SCAN_CODE:          ScanCmd(); break;
        case CLEAR_CODE:         ClearCmd(); break;
        default:                 executed = false; break;
    }
    return executed;
}

boolean IClass::ForwardSearch (const char* string) {
    Regexp re(string);
    boolean successful = false;

    if (
        _text->ForwardSearch(&re, _editor->Dot()) >= 0
        || _text->ForwardSearch(&re, _text->BeginningOfText()) >= 0
    ) {
        _editor->Select(re.EndOfMatch(), re.BeginningOfMatch());
        successful = true;
    }
    return successful;
}

boolean IClass::BackwardSearch (const char* string) {
    Regexp re(string);
    boolean successful = false;

    if (
        _text->BackwardSearch(&re, _editor->Mark()) >= 0
        || _text->BackwardSearch(&re, _text->EndOfText()) >= 0
    ) {
        _editor->Select(re.EndOfMatch(), re.BeginningOfMatch());
        successful = true;
    }
    return successful;
}

void IClass::UnselectBrowsers () {
    int index;

    if ((index = _classes->Selection()) >= 0) {
        _classes->UnselectAll();

    } else if ((index = _parents->Selection()) >= 0) {
        _parents->UnselectAll();

    } else if ((index = _children->Selection()) >= 0) {
        _children->UnselectAll();
    }
}

void IClass::Update () {
    int value;
    _state->GetValue(value);

    if (value == '\r') {
        UpdateCurClass();
        UpdateParentBrowser();
        UpdateChildBrowser();

    } else if (value == '\t' || value == '\033') {
        SelectCurClass();
        _focus = _curClass;

    } else if (value == '\007') {
        Complain();
    }

    if (value != QUIT_CODE) {
        UnselectBrowsers();
        _state->SetValue('\0');
    }
}

void IClass::UpdateCurClass () {
    int index;

    if ((index = _classes->Selection()) >= 0) {
        UpdateCurClass(_classes->String(index));

    } else if ((index = _parents->Selection()) >= 0) {
        UpdateCurClass(_parents->String(index));

    } else if ((index = _children->Selection()) >= 0) {
        UpdateCurClass(_children->String(index));

    } else {
        UpdateCurClass(_curClass->Text());
    }
}

void IClass::UpdateCurClass (const char* string) {
    if (strcmp(string, NONE) != 0) {
        if (strcmp(string, _curClass->Text()) != 0) {
            _curClass->Message(string);
        }
        ClassInfo* info = _cbuf->Info(string);

        if (info == nil) {
            SelectCurClass();
            Complain();

        } else {
            Visit(info->Path());
            int line = info->LineNumber();
	    _editor->Edit(_text, _text->LineIndex(line));
            ForwardSearch(info->Name());
            boolean backward_success = BackwardSearch("//: ");
	    _editor->ScrollToSelection();
            if (backward_success) ForwardSearch(info->Name());
            _focus = _editor;
        }
    }
}

void IClass::UpdateClassesBrowser () {
    const char* className;
    int i = 0;

    _classes->Clear();

    int count = 0;
    for (;;) {
        className = _cbuf->Class(i++);
        if (className == nil) {
            break;
        } else {
            _classes->Append(className);
            ++count;
        }
    }
    if (_completions != nil) {
        delete _completions;
    }
    _completions = new const char*[count];
    for (i = 0; i < count; ++i) {
        _completions[i] = _cbuf->Class(i);
    }
    _curClass->Completions(_completions, count);
}

void IClass::UpdateParentBrowser () {
    const char* parent;
    int i = 0;

    _parents->Clear();

    for (;;) {
        parent = _cbuf->Parent(_curClass->Text(), i++);

        if (parent == nil) {
            if (i == 1) {
                _parents->Append(NONE);
            }
            break;
        } else {
            _parents->Append(parent);
        }
    }
}

void IClass::UpdateChildBrowser () {
    const char* child;
    int i = 0;

    _children->Clear();

    for (;;) {
        child = _cbuf->Child(_curClass->Text(), i++);
        if (child == nil) {
            if (i == 1) {
                _children->Append(NONE);
            }
            break;
        } else {
            _children->Append(child);
        }
    }
}

void IClass::SelectCurClass () {
    _curClass->Select(0, strlen(_curClass->Text()));
}

void IClass::UnselectCurClass () {
    _curClass->Select(strlen(_curClass->Text()));
}

Interactor* IClass::Interior () {
    int space = Math::round(fspace*ivcm);
    const char* showButton = GetAttribute("showButton");
    Interactor* selector = SelectedClass();

    if (showButton != nil && strcmp(showButton, "true") == 0) {
	selector = new HBox(
	    selector,
	    new HGlue(space, space, 0),
	    new PushButton(" Show ", _state, '\r')
	);
    }

    return new VBox(
        new HBox(Commands(), new HGlue),
        new HBorder,
        new MarginFrame(
            new VBox(
                new HBox(
                    Classes(),
                    new HGlue(space, space, 0),
                    new VBox(
			selector,
                        new VGlue(space, space, 0),
                        Parents(),
                        new VGlue(space, space, 0),
                        Children()
                    )
                ),
                new VGlue(space, space, 0),
                Editor()
            ), space, space, 0
        )
    );
}

static MenuInfo fileMenu[] = {
    { "Scan Files/Directories...", SCAN_LBL, SCAN_CODE },
    { "Clear Classes", CLEAR_LBL, CLEAR_CODE },
    { "Quit", QUIT_LBL, QUIT_CODE },
    { nil }
};

static MenuInfo searchMenu[] = {
    { "Forward Search...", FSRCH_LBL, FSRCH_CODE },
    { "Backward Search...", BSRCH_LBL, BSRCH_CODE },
    { "Go to Line...", GOTO_LBL, GOTO_CODE },
    { nil }
};

static PulldownMenu* MakePulldown(
    const char* name, MenuInfo* item, IClass* ic
) {
    PulldownMenu* menu = new PulldownMenu(
	new Message(name, Left, Math::round(.1*ivcm))
    );

    for (MenuInfo* i = item; i->_lbl != nil; i++) {
        menu->Include(new Command(i->_lbl, i->_klbl, i->_kcode, ic));
    }
    return menu;
}

Interactor* IClass::Commands () {
    MenuBar* menuBar = new MenuBar;

    menuBar->Include(MakePulldown("File", fileMenu, this));
    menuBar->Include(MakePulldown("Search", searchMenu, this));

    return menuBar;
}

Interactor* IClass::Classes () {
    return new VBox(
        new HBox(
            new Message("classes"),
            new HGlue
        ),
        new Frame(
            AddScroller(_classes)
        )
    );
}

Interactor* IClass::SelectedClass () {
    return new VBox(
        new HBox(
            new Message("selected class"),
            new HGlue
        ),
        new Frame(new MarginFrame(_curClass, 2))
    );            
}

Interactor* IClass::Parents () {
    return new VBox(
        new HBox(
            new Message("parents"),
            new HGlue
        ),
        new Frame(AddScroller(_parents))
    );
}

Interactor* IClass::Children () {
    return new VBox(
        new HBox(
            new Message("children"),
            new HGlue
        ),
        new Frame(AddScroller(_children))
    );
}

Interactor* IClass::Editor () {
    HBox* fileBox = new HBox(
        new Message("file: "),
        _fileIndic,
        new HGlue
    );
    fileBox->Propagate(false);

    return new VBox(
        fileBox,
        new Frame(AddScroller(_editor))
    );
}

Interactor* IClass::AddScroller (Interactor* i) {
    return new HBox(
        new MarginFrame(i, 2),
        new VBorder,
	new VScrollBar(i)
    );
}

void IClass::InsertDialog (Interactor* dialog) {
    World* world = GetWorld();

    Coord x, y;
    Align(Center, 0, 0, x, y);
    GetRelative(x, y, world);

    world->InsertTransient(dialog, this, x, y, Center);
}

void IClass::RemoveDialog (Interactor* dialog) {
    GetWorld()->Remove(dialog);
}

void IClass::Complain (const char* msg) {
    if (msg == nil) {
        fprintf(stderr, "%c", '\007');

    } else {
        AcknowledgeDialog complaint(msg);

        InsertDialog(&complaint);
        complaint.Acknowledge();
        RemoveDialog(&complaint);
    }
}

void IClass::ScanCmd () {
    InsertDialog(_scanner);

    if (_scanner->Accept()) {
        FileBrowser* browser = _scanner->GetBrowser();

        for (int i = 0; i < browser->Selections(); ++i) {
            int index = browser->Selection(i);
            _cbuf->Search(browser->Path(index));
        }
    }
    RemoveDialog(_scanner);
    UpdateClassesBrowser();
}

void IClass::ClearCmd () {
    ConfirmDialog dialog("Really clear all classes?");
    InsertDialog(&dialog);
    boolean accepted = dialog.Confirm() == 'y';
    RemoveDialog(&dialog);

    if (accepted) {
        World* world = GetWorld();
        const char* recursive = world->GetAttribute("recursive");
        const char* verbose = world->GetAttribute("verbose");
        const char* CPlusPlusFiles = world->GetAttribute("CPlusPlusFiles");

        delete _cbuf;
        _cbuf = new ClassBuffer(
            strcmp(recursive, "true") == 0, strcmp(verbose, "true") == 0,
            strcmp(CPlusPlusFiles, "true") == 0
        );

        UpdateClassesBrowser();
        UpdateParentBrowser();
        UpdateChildBrowser();
    }
}

void IClass::QuitCmd () { _state->SetValue(QUIT_CODE); }

void IClass::ForwardSearchCmd () {
    if (_fwdSearch == nil) {
        _fwdSearch = new StringDialog("Search forward for: ", Math::round(3*ivinches));
    }

    InsertDialog(_fwdSearch);
    
    if (_fwdSearch->Accept()) {
        if (ForwardSearch(_fwdSearch->String())) {
            _editor->ScrollToSelection();
        } else {
            Complain();
        }
    }
    RemoveDialog(_fwdSearch);
}

void IClass::BackwardSearchCmd () {
    if (_bwdSearch == nil) {
        _bwdSearch = new StringDialog("Search backward for:", Math::round(3*ivinches));
    }

    InsertDialog(_bwdSearch);
    
    if (_bwdSearch->Accept()) {
        if (BackwardSearch(_bwdSearch->String())) {
            _editor->ScrollToSelection();
        } else {
            Complain();
        }
    }
    RemoveDialog(_bwdSearch);
}

void IClass::GotoCmd () {
    if (_goto == nil) {
        int width = output->GetFont()->Width("9999999");
        _goto = new StringDialog("Go to line:", width);
    }

    InsertDialog(_goto);
    
    if (_goto->Accept()) {
        const char* string = _goto->String();
        int line = atoi(string);
        _editor->Select(_text->LineIndex(line-1));
        _editor->ScrollToSelection();
    }
    RemoveDialog(_goto);
}
